[ResourceMonitor] Send ContentRuleList of Checker from UIProcess to WebProcess.

basuke opened the pull request:
#### a097c5ab60cdccd9f066cf7c44a69d88cb14b60c
<pre>
[ResourceMonitor] Send ContentRuleList of Checker from UIProcess to WebProcess.
<a href="https://bugs.webkit.org/show_bug.cgi?id=284694">https://bugs.webkit.org/show_bug.cgi?id=284694</a>
<a href="https://rdar.apple.com/141645827">rdar://141645827</a>

Reviewed by Ben Nham.

Every web content process has its own checker. They share same content rule list which is mapped file to be
shared among the processes. We need to send the list from UIProcess that initialize the rule list to web
content process where ResourceMonitorChecker is running in it.

In this patch, actual preparation of rule list is not included.

* Source/WebCore/Headers.cmake:
* Source/WebCore/loader/ResourceMonitorChecker.h:
* Source/WebKit/UIProcess/Cocoa/WebProcessPoolCocoa.mm:
(WebKit::WebProcessPool::platformLoadResourceMonitorRuleList):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::loadRequest):
(WebKit::WebPageProxy::loadRequestWithNavigationShared):
* Source/WebKit/UIProcess/WebProcessPool.cpp:
(WebKit::WebProcessPool::cachedResourceMonitorRuleList):
(WebKit::WebProcessPool::loadOrUpdateResourceMonitorRuleList):
(WebKit::WebProcessPool::platformLoadResourceMonitorRuleList):
* Source/WebKit/UIProcess/WebProcessPool.h:
* Source/WebKit/UIProcess/WebProcessProxy.cpp:
(WebKit::WebProcessProxy::requestResourceMonitorRuleLists):
(WebKit::WebProcessProxy::setResourceMonitorRuleListsIfRequired):
* Source/WebKit/UIProcess/WebProcessProxy.h:
* Source/WebKit/WebProcess/WebProcess.cpp:
(WebKit::WebProcess::setResourceMonitorContentRuleList):
* Source/WebKit/WebProcess/WebProcess.h:
* Source/WebKit/WebProcess/WebProcess.messages.in:

Canonical link: <a href="https://commits.webkit.org/288017@main">https://commits.webkit.org/288017@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7090a23c3b6219b8273e7eff5c090fc6275c54bb

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/81689 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/1214 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/35638 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/86230 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/32683 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/83795 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/1245 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/9033 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/63736 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/21461 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/84759 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/901 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/74341 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/44021 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/800 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/28515 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/31141 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/72190 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/29117 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/87673 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/8926 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/6335 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/72071 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/9110 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/70160 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/71301 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/17753 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/15381 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/14297 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/8878 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/14412 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/8718 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/12245 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/10527 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

bnham commented on Source/WebKit/UIProcess/Cocoa/WebProcessPoolCocoa.mm:
```diff
@@ -1333,4 +1333,11 @@ static void webProcessPoolHighDynamicRangeDidChangeCallback(CFNotificationCenter
+#if ENABLE(CONTENT_EXTENSIONS)
+void WebProcessPool::platformLoadResourceMonitorRuleList(CompletionHandler<void()>&& completionHandler)
+{
+    completionHandler();
```
I think this part of the patch is missing? It seems like it is supposed to set `m_resourceMonitorRuleListCache`.

bnham replied:
Talked to Basuke, he said he would implement this in a future patch.

bnham commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -1997,6 +1997,15 @@ RefPtr<API::Navigation> WebPageProxy::loadRequest(WebCore::ResourceRequest&& req
+    if (preferences().iFrameResourceMonitoringEnabled()) {
+        Ref processPool = m_legacyMainFrameProcess->processPool();
+        if (RefPtr ruleList = processPool->cachedResourceMonitorRuleList())
+            protectedLegacyMainFrameProcess()->send(Messages::WebProcess::SetResourceMonitorContentRuleList(ruleList->data()), 0);
```
Seems like this should be only sent if the process hasn't already gotten the list in the past, rather than on every single `loadRequest` call. I would also put something like that in `loadRequestWithNavigationShared` instead since that is the common chokepoint for request loading.

basuke replied:
I'll move back the initialization to webprocessproxy to remember the status. And from `loadRequestWithNavigationShared` to send a request to process proxy to load that.

bnham commented on Source/WebKit/UIProcess/WebProcessPool.cpp:
```diff
@@ -2681,4 +2685,49 @@ void WebProcessPool::updateWebProcessSuspensionDelayWithPacing(WeakHashSet<WebPr
 
+#if ENABLE(CONTENT_EXTENSIONS)
+
+constexpr static Seconds resourceMonitorRuleListCheckInterval = 24_h;
```
If you are going to have `loadRequestWithNavigationShared` call `cachedResourceMonitorRuleList`, then you could also consider a lazier strategy that doesn't involve a timer would be to just store the timestamp at which you last refreshed the rule list. If that timestamp was more than 24 hours ago, then perform a refresh. See `approximateContinuousTimeNanoseconds` (although that only works on Darwin, on other platforms, maybe just use a WallTime).

Whether you use a timer or just store a timestamp, I would think about maybe adding some randomness to the check interval (like 24 hr + 1 hour of uniform randomness) since this can involve talking to a server, and we want to try to prevent thundering herds.

basuke replied:
I've started with that lazy implementation with expiration. But that doesn't cover the case that one and only tab is running for more than 24 hours. If no request happens, list won't be updated. Well, no request, no new activity so this shouldn't be big deal, but covering these in timer may be straightforward.

Adding randomness is reasonable, but the timer began based on the timing of user launch and we don't need to think about the server access, do we?

bnham commented on Source/WebKit/UIProcess/WebProcessPool.cpp:
```diff
@@ -2681,4 +2685,49 @@ void WebProcessPool::updateWebProcessSuspensionDelayWithPacing(WeakHashSet<WebPr
+
+void WebProcessPool::loadOrUpdateResourceMonitorRuleList()
+{
+    WEBPROCESSPOOL_RELEASE_LOG(Process, "createWebPage: delaying WebProcess launch until first load");
```
This log line isn't correct.

bnham commented on Source/WebKit/WebProcess/WebProcess.cpp:
```diff
@@ -2471,6 +2476,22 @@ Ref<WebCookieJar> WebProcess::protectedCookieJar()
+#if ENABLE(CONTENT_EXTENSIONS)
+void WebProcess::setResourceMonitorContentRuleList(WebCompiledContentRuleListData&& ruleListData)
+{
+    RefPtr compiledContentRuleList = WebCompiledContentRuleList::create(WTFMove(ruleListData));
```
How long does this take? It will take up CPU time on the main thread of WebProcess, potentially in the middle of a page load.

basuke replied:
It's just copying trivial data and very cheap.

bnham replied:
This particular line of code seems cheap but not sure about the `addContentExtension` line below.

basuke replied:
Hmm, sure. I double check this.

basuke replied:
Checked with Alex and he confirmed this is very cheap and no problem using it in main thread.

bnham reviewed (changes requested):
We should not send the message on each `loadRequest`; it should only be sent if hasn't been sent to the process before. Also, it looks like `m_resourceMonitorRuleListCache` isn't set anywhere.

bnham commented on Source/WebKit/WebProcess/WebProcess.messages.in:
```diff
@@ -229,4 +229,8 @@ messages -> WebProcess : AuxiliaryProcess WantsAsyncDispatchMessage {
 #endif
+
+#if ENABLE(CONTENT_EXTENSIONS)
+    SetResourceMonitorContentRuleList(WebKit::WebCompiledContentRuleListData ruleListData)
```
Please add `[DeferSendingIfSuspended]` to this IPC. If the receiver is suspended for a long time (like multiple days), we don't want multiple copies of this message building up in the receiver's message queue.

bnham commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -1997,6 +1997,7 @@ RefPtr<API::Navigation> WebPageProxy::loadRequest(WebCore::ResourceRequest&& req
 #endif
 
     loadRequestWithNavigationShared(protectedLegacyMainFrameProcess(), m_webPageID, navigation, WTFMove(request), shouldOpenExternalURLsPolicy, isPerformingHTTPFallback, userData, ShouldTreatAsContinuingLoad::No, isNavigatingToAppBoundDomain(), std::nullopt, std::nullopt);
+
```
nit: unnecessary newline

bnham commented on Source/WebKit/UIProcess/WebProcessPool.cpp:
```diff
@@ -2681,4 +2686,58 @@ void WebProcessPool::updateWebProcessSuspensionDelayWithPacing(WeakHashSet<WebPr
+
+        protectedThis->m_resourceMonitorRuleListLoading = false;
+
+        RefPtr ruleList = protectedThis->m_resourceMonitorRuleListCache;
```
On non-Darwin platforms, `platformLoadResourceMonitorRuleList` always returns nullptr. This means that every time `cachedResourceMonitorRuleList` runs, it does some unnecessary work to try to load an empty rule list.

It might be nice to optimize this out, e.g. by making sure that you only call platformLoadResourceMonitorRuleList or loadOrUpdateResourceMonitorRuleList at most once a day.

bnham commented on Source/WebKit/UIProcess/WebProcessProxy.h:
```diff
@@ -829,6 +835,12 @@ class WebProcessProxy final : public AuxiliaryProcessProxy {
+
+#if ENABLE(CONTENT_EXTENSIONS)
+    bool m_resourceMonitorRuleListRequestedBySomePage { false };
+    RefPtr<WebCompiledContentRuleList> m_resourceMonitorRuleListSentTo;
```
I would just name this `m_resourceMonitorRuleList`. Or just make it `bool m_hasSentResourceMonitorRuleList`.

basuke replied:
Need to update the list when new list is delivered, so comparison is required. I'll change the name as suggested.

basuke committed:
[ResourceMonitor] Send ContentRuleList of Checker from UIProcess to WebProcess.
https://bugs.webkit.org/show_bug.cgi?id=284694
rdar://141645827

Reviewed by Ben Nham.

Every web content process has its own checker. They share same content rule list which is mapped file to be
shared among the processes. We need to send the list from UIProcess that initialize the rule list to web
content process where ResourceMonitorChecker is running in it.

In this patch, actual preparation of rule list is not included.

* Source/WebCore/Headers.cmake:
* Source/WebCore/loader/ResourceMonitorChecker.h:
* Source/WebKit/UIProcess/Cocoa/WebProcessPoolCocoa.mm:
(WebKit::WebProcessPool::platformLoadResourceMonitorRuleList):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::loadRequest):
(WebKit::WebPageProxy::loadRequestWithNavigationShared):
* Source/WebKit/UIProcess/WebProcessPool.cpp:
(WebKit::WebProcessPool::cachedResourceMonitorRuleList):
(WebKit::WebProcessPool::loadOrUpdateResourceMonitorRuleList):
(WebKit::WebProcessPool::platformLoadResourceMonitorRuleList):
* Source/WebKit/UIProcess/WebProcessPool.h:
* Source/WebKit/UIProcess/WebProcessProxy.cpp:
(WebKit::WebProcessProxy::requestResourceMonitorRuleLists):
(WebKit::WebProcessProxy::setResourceMonitorRuleListsIfRequired):
* Source/WebKit/UIProcess/WebProcessProxy.h:
* Source/WebKit/WebProcess/WebProcess.cpp:
(WebKit::WebProcess::setResourceMonitorContentRuleList):
* Source/WebKit/WebProcess/WebProcess.h:
* Source/WebKit/WebProcess/WebProcess.messages.in:

Canonical link: https://commits.webkit.org/288017@main
